### PR TITLE
rgw: support obj cryption stack on compression

### PR DESCRIPTION
**Pull request by Zhiwei-Dai**
modified:   src/rgw/rgw_compression.cc
	modified:   src/rgw/rgw_compression.h
	modified:   src/rgw/rgw_op.cc

Fix: https://tracker.ceph.com/issues/19988

Signed-off-by: Dai Zhiwei daizhiwei3@huawei.com
Signed-off-by: luo rixin luorixin@huawei.com

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_compression.cc [hunk 1]
```diff
@@ -67,7 +67,23 @@ int RGWPutObj_Compress::process(bufferlist&& in, uint64_t logical_offset)
     }
     // end of compression stuff
   }
-  return Pipe::process(std::move(out), logical_offset);
+
+  if (out.length()) {
+    in_len = in.length();
+    out_len = in.length();
```
Comment (rosinL) [inline src/rgw/rgw_compression.cc]: out_len = in.length()?

File: src/rgw/rgw_compression.cc [hunk 2]
```diff
@@ -67,7 +67,23 @@ int RGWPutObj_Compress::process(bufferlist&& in, uint64_t logical_offset)
...
+      logical_offset = logical_offset + out_len - in_len;
+    }
+    compressed_ofs = logical_offset;
+  } else {
+    compressed_ofs = logical_offset + out_len - in_len;
+    in_len = 0;
+    out_len = 0;
+  }
+
+  return Pipe::process(std::move(out), compressed_ofs);
```
Comment (rosinL) [inline src/rgw/rgw_compression.cc]: cryption and compression, which is first?  How to know the data are both cryptoed and compressed? Is it an option to user to choose both crpyto and compress

File: src/rgw/rgw_compression.cc (hunk 1 above)
Comment (wade-wu) [reply src/rgw/rgw_compression.cc]: .length()是否会计算？out_len = in_len

Comment (Zhiwei-Dai) [reply src/rgw/rgw_compression.cc]: > out_len = in.length()?

oops, a mistake.

File: src/rgw/rgw_compression.cc (hunk 2 above)
Comment (Zhiwei-Dai) [reply src/rgw/rgw_compression.cc]: 当前只支持：put过程是先压缩，再加密；get是先解密，再解压。压缩与否是在指定placement的属性里决定的；而加密是由客户端s3接口以及config文件中指定的加密算法一起决定使能与否的。

**Pull request closed without merging**